Data/Aeson/Types/Class.hs

```diff
@@ -250,6 +257,15 @@ class KeyValue kv where
     (.=) :: ToJSON v => Text -> v -> kv
     infixr 8 .=
 
+data ToJSONKeyFunction a
+    = ToJSONKeyText (a -> Text, a -> Encoding)
+    | ToJSONKeyValue (a -> Value, a -> Encoding)
```

Comment:
I thought about this, but I feel like the dependency on `contravariant` probably not worth it when we could just provide a specialized version of the function instead.